### PR TITLE
Int 35 show error banner preventing submit

### PR DESCRIPTION
**Pull request by jgedeon-osi**
### Jira Story

- [Name of Story INT-35](https://osi-cwds.atlassian.net/browse/INT-35)

### Purpose

Give better error reporting to users on failure to submit referral

### Background

This story adds two things, 

- a top level banner that shows a count of errors
- an error detail card that shows up at the top

### Notes for Reviewer

This PR only adds the first item, the second is to come in a later PR.
The intake api also needs to be updated to report errors on submission of referral in the same format as ferb.

### CODE REVIEW

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 1]
```diff
@@ -1,6 +1,6 @@
-### Pivotal Story
+### Jira Story
 
-- [Name of Story NNNNN](https://www.pivotaltracker.com/story/show/NNNNN)
+- [Name of Story INT-NNN](https://osi-cwds.atlassian.net/browse/INT-NNN)
```
Comment (saraOSI) [inline .github/PULL_REQUEST_TEMPLATE.md]: What if we're not working out of INT?!

File: app/javascript/common/App.jsx [hunk 2]
```diff
@@ -23,10 +27,12 @@ export class App extends React.Component {
 App.propTypes = {
   actions: PropTypes.object.isRequired,
   children: PropTypes.object.isRequired,
-  remoteError: PropTypes.object,
+  errorCount: PropTypes.number,
+  hasGenericError: PropTypes.bool,
 }
 const mapStateToProps = (state, _ownProps) => ({
```
Comment (saraOSI) [inline app/javascript/common/App.jsx]: We should really norm on whether or not to include `_ownProps` in situations like this. It's a second param, so it's not needed. And maybe by now we all know that `ownProps` is the second param that `mapStateToProps` can take? I'd probably vote for removing this, but I'm also ok with leaving it.

File: app/javascript/common/PageError.jsx [hunk 3]
```diff
@@ -1,13 +1,21 @@
...
 import React from 'react'
 
-const PageError = () => (
+const PageError = ({errorCount}) => (
   <div className='page-error'>
     <div className='container'>
       <div className='row'>
-        <p className='text-center'>Something went wrong, sorry! Please try your last action again.</p>
+        <p className='text-center'>
+          {!errorCount && 'Something went wrong, sorry! Please try your last action again.'}
```
Comment (saraOSI) [inline app/javascript/common/PageError.jsx]: I know it throws off not having an explicit return, but maybe you could create a local param here that is `hasErrorCount`, and then use that in your render checks? eg: 
```
const hasErrorCount = Boolean(errorCount)
{hasErrorCount && "blah"}
{hasErrorCount && "woohoo"}
```

Might also make it more clear that one message is to be used when the errors have no count, vs the message for when there is a specific number of errors.

File: .github/PULL_REQUEST_TEMPLATE.md (hunk 1 above)
Comment (jgedeon-osi) [reply .github/PULL_REQUEST_TEMPLATE.md]: We change it manually? Or we change the template when that happens :)

File: app/javascript/common/App.jsx (hunk 2 above)
Comment (jgedeon-osi) [reply app/javascript/common/App.jsx]: I agree we should norm on this. I could go either way so I'll leave it for now.

File: app/javascript/common/PageError.jsx (hunk 3 above)
Comment (jgedeon-osi) [reply app/javascript/common/PageError.jsx]: I agree, I've fixed this.

File: app/javascript/selectors/errorsSelectors.js [hunk 4]
```diff
@@ -0,0 +1,8 @@
+import {Map} from 'immutable'
+import {createSelector} from 'reselect'
+
+export const getGenericErrors = (state) => state.getIn(['errors', 'unknown'], Map())
+export const getHasGenericErrorValueSelector = createSelector(
```
Comment (AmanShurafa) [inline app/javascript/selectors/errorsSelectors.js]: The name is ridiculously long. Not much to do about it though. Can we at least remove `Value`?

File: app/javascript/common/PageError.jsx (hunk 3 above)
Comment (AmanShurafa) [reply app/javascript/common/PageError.jsx]: I miss Ruby

File: app/javascript/selectors/errorsSelectors.js [hunk 5]
```diff
@@ -1,8 +1,19 @@
...
 
-export const getGenericErrors = (state) => state.get('errors', Map())
+export const getErrors = (state) => state.get('errors', Map())
 export const getHasGenericErrorValueSelector = createSelector(
-  getGenericErrors,
-  (genericErrors) => (genericErrors.size > 0)
+  getErrors,
+  (errors) => (errors.size > 0)
+)
+export const getTotalScreeningSubmissionErrorValueSelector = createSelector(
```
Comment (AmanShurafa) [inline app/javascript/selectors/errorsSelectors.js]: How about something shorter like: `getTotalSubmitScreeningErrorSelector`

File: spec/features/errors_spec.rb [hunk 6]
```diff
@@ -58,35 +58,69 @@
...
-        :get, intake_api_url(ExternalRoutes.intake_api_screening_path(screening.id))
-      ).and_return(json_body(screening.to_json, status: 200))
-      stub_empty_relationships_for_screening(screening)
-      stub_empty_history_for_screening(screening)
-      Feature.run_with_activated(:referral_submit) do
-        example.run
+    context 'with referral_submit' do
+      let(:referral_id) { FFaker::Guid.guid }
+
+      scenario 'hide the error banner after a successful action' do
```
[the remark this answers is not in the document]
Comment (jgedeon-osi) [reply spec/features/errors_spec.rb]: Renamed.

File: spec/features/errors_spec.rb [hunk 7]
```diff
@@ -58,35 +58,69 @@
...
       stub_request(
-        :post,
-        intake_api_url(ExternalRoutes.intake_api_screening_submit_path(screening.id))
-      ).and_return(json_body([].to_json, status: 200))
-      click_button 'Submit'
+        :put, intake_api_url(ExternalRoutes.intake_api_screening_path(screening.id))
+      ).and_return(json_body(screening.to_json, status: 200))
+
+      # clicking the save button but for some reason xvfb shows the banner over the button no matter
+      # what when using the standard methods of clicking
```
[the remark this answers is not in the document]
Comment (jgedeon-osi) [reply spec/features/errors_spec.rb]: No bug, not sure where or how this would be fixed.

File: app/javascript/selectors/errorsSelectors.js (hunk 5 above)
Comment (jgedeon-osi) [reply app/javascript/selectors/errorsSelectors.js]: We name all our selectors that return a value with ValueSelector on the end. Not sure what I could remove before it became confusing.

File: app/javascript/selectors/errorsSelectors.js (hunk 4 above)
Comment (jgedeon-osi) [reply app/javascript/selectors/errorsSelectors.js]: We name all selectors that return a value with ValueSelector on the end so not really.

**Pull request merged**